File: Assets/Resources/LuaScripts/Framework/MVC/UIMediator.lua
--[[********************************************************************
*		作者： jordenwu
*		时间： 2018-07-04
*		描述： 模块UI中间件
*********************************************************************--]]
local UIMediatorClass = DeclareClass("UIMediatorClass")
local null = {}

function UIMediatorClass:ctor()
    
    self._module = false
    self._forms = {}
    self._belongUIStateName = false
    self._inState = false
end

function UIMediatorClass:Create(md)
    self._module = md
    self._belongUIStateName = self:vGetBelongUIStateName() or false

    self:vOnInitialize()
end

function UIMediatorClass:Destroy()
    self:vOnUninitialize()
    self._module = false

    for i = 1, #self._forms do
        if self._forms[i] ~= null then
            LogE("UIMediatorClass.Destroy : FormClass %s of %s is living", self._forms[i].__classname, self.__classname)
        end
    end
    
    self._forms = {}
    self._belongUIStateName = false
    self._inState = false
end

--UIState改变处理
function UIMediatorClass:ChangeUIState(callType, changeType, oldStateName, newStateName, stateParam)
    
    if not self._belongUIStateName then
        return
    end

    local old = false
    local new = false
    for i = 1, #self._belongUIStateName do
        local name = self._belongUIStateName[i]

        if oldStateName and oldStateName == name then
            old = true
        end

        if newStateName and newStateName == name then
            new = true
        end

        if old and new then
            break
        end
    end

    if old ~= new then
        if old then
            if callType == 1 then
                self._inState = false
                self:vOnUIStateOut(oldStateName, newStateName, changeType)
           end
        else
            if callType == 2 then
                self._inState = true
                self:vOnUIStateIn(oldStateName, newStateName, stateParam, changeType)
            end
        end
    elseif old and new then
        if callType == 0 then
            self:vOnUIStateStay(oldStateName, newStateName, stateParam, changeType)
        end
    end
end

-----------
function UIMediatorClass:vGetBelongUIStateName()
end


function UIMediatorClass:vOnInitialize()
end


function UIMediatorClass:vOnUninitialize()
end


function UIMediatorClass:vOnUIStateIn(oldStateName, newStateName, stateParam, changeType)
end


function UIMediatorClass:vOnUIStateOut(oldStateName, newStateName, changeType)
end


function UIMediatorClass:vOnUIStateStay(oldStateName, newStateName, stateParam, changeType)
end

function UIMediatorClass:vOnUpdateUI(id, argument)
end

--[[
    @desc: 操作回调
    --@id:操作标识，推荐使用字符串和数字
	--@argument: 操作参数
    @return:是否阻断向上传递操作
]]
function UIMediatorClass:vOnAction(id, argument)
end


function UIMediatorClass:IsInState()
    return self._inState
end


--[[
    @desc: 创建窗口实例
    --@cls:类名
	--@argument:参数
	--@customID: 自定义ID
    @return:
]]
function UIMediatorClass:CreateFormClass(cls, argument, customID)
    if not cls then
        LogE("UIMediatorClass.CreatePrefabClass : invalid parameter check Class Required ?")
        return nil
    end

    local obj = cls.new()
    obj:Create(self, argument, customID)
    self._forms[#self._forms+1] = obj
    return obj
end


--[[
    @desc: 销毁窗口类
    --@obj: 实例
    @return:无
]]
function UIMediatorClass:DestroyFormClass(obj)
    if not obj then
        LogE("UIMediatorClass.DestroyFormClass : invalid parameter")
        return
    end

    for i = 1, #self._forms do
        if self._forms[i] == obj then
            self._forms[i] = null
            break
        end
    end

    obj:Destroy()
end


--[[
    @desc: 通过ID销毁窗口实例
    --@customID: 自定义ID
    @return:
]]
function UIMediatorClass:DestroyFormClassByCustomID(customID)
    if not customID then
        LogE("UIMediatorClass.DestroyFormClassByCustomID : invalid parameter")
        return
    end
    
    for i = #self._forms, 1, -1 do
        local pc = self._forms[i]
        if pc ~= null and pc._customID == customID then
            self._forms[i] = null
            pc:Destroy()
        end
    end

end

--[[
    @desc: 销毁所有
    @return:无
]]
function UIMediatorClass:DestroyAllFormClass()
    for i = 1, #self._forms do
        local obj = self._forms[i]
        if obj ~= null then
            self._forms[i] = null
            obj:Destroy()
        end        
    end
end

--[[
    @desc: 发送更新UI（只能在实例内部调用)
    --@id:刷新ID标识
	--@argument:参数
	--@dummy: 不要管这个参数
    @return:
]]
function UIMediatorClass:UpdateUI(id, argument, dummy)
    if not id then
        LogE("UIMediatorClass.UpdateUI : invalid parameter")
        return
    end

    if dummy then
        if self:vOnUpdateUI(id, argument) then
            return
        end
    end
    
    for i = #self._forms, 1, -1 do
        local obj = self._forms[i]
        if obj ~= null then
            obj:vOnUpdateUI(id, argument, true)
        end
    end
end

--[[
    @desc: 发送操作
    --@id:操作ID标识
	--@argument:参数
	--@dummy: 不要管这个参数
    @return:
]]
function UIMediatorClass:Action(id, argument, dummy)
    if not id then
        LogE("UIMediatorClass.SendAction : invalid parameter")
        return
    end

    if dummy then
        if self:vOnAction(id, argument) then
            return
        end
    end
    
    self._module:vOnAction(id, argument)
end

File: Assets/Scripts/Framework/MVC/UIMediator.cs
using System;
using JW.Common;
using JW.Framework.Asset;

namespace JW.Framework.MVC
{
    public abstract class UIMediator
    {
        protected ModuleBase Module;

        // 管理的窗口列表
        private JWArrayList<UIFormClass> _forms;

        // 是否在UIMediator所属的状态中
        private bool _inState;

#region 外部不要调用
        /// <summary>
        /// 创建，外部不要调用
        /// </summary>
        public void Create(ModuleBase module)
        {
            Module = module;

            _forms = new JWArrayList<UIFormClass>(1);
            _inState = false;
            //注入到UI状态服务
            UIStateService.GetInstance().AddMediator(this, GetBelongsUIStateName());

            OnInitialize();
        }

        /// <summary>
        /// 销毁对象，外部不要调用
        /// </summary>
        public void Destroy()
        {
            OnUninitialize();

            UIStateService.GetInstance().RemoveMediator(this);

            for (int i = 0; i < _forms.Count; i++)
            {
                if (_forms[i] != null)
                {
                    JW.Common.Log.LogE("UIMediator.Destroy : FormClass {0} of {1} is living", _forms[i].GetType().FullName, GetType().FullName);
                }
            }
            _inState = false;
            _forms.Clear();
            Module = null;
        }
    
        //延迟清理
        public void ProcessControl()
        {
            for (int i = _forms.Count - 1; i >= 0; --i)
            {
                if (_forms[i] == null)
                {
                    _forms.RemoveAt(i);
                    continue;
                }
            }
        }

        /// UIStateService UI状态回掉，外部不要调用
        public void ChangeState(int changeType, bool inState, string oldStateName, string newStateName, object stateParam)
        {
            if (inState)
            {
                _inState = true;
                OnUIStateIn(changeType, oldStateName, newStateName, stateParam);
            }
            else
            {
                _inState = false;
                OnUIStateOut(changeType, oldStateName, newStateName);
            }
        }

        /// UIStateService 刷新状态，外部不要调用
        public void RefreshState(int changeType, string oldStateName, string newStateName, object stateParam)
        {
            OnUIStateStay(changeType, oldStateName, newStateName, stateParam);
        }

#endregion

        /// <summary>
        /// 更新UI
        /// </summary>
        /// <param name="id">更新ID标识</param>
        /// <param name="param">更新参数</param>
        public void UpdateUI(string id, object param)
        {
            if (OnUpdateUI(id, param))
            {
                return;
            }

            for (int i = _forms.Count - 1; i >= 0; --i)
            {
                if (_forms[i] != null)
                {
                    _forms[i].UpdateUI(id, param);
                }
            }
        }

        /// <summary>
        /// 操作
        /// </summary>
        /// <param name="id">操作ID标识</param>
        /// <param name="param">操作参数</param>
        /// <param name="sponsor">是否是操作发起者，调用时可以不用管这个参数</param>
        public void Action(string id, object param, bool sponsor = true)
        {
            if (!sponsor)
            {
                if (OnAction(id, param))
                {
                    return;
                }
            }

            if (Module != null)
            {
                Module.Action(id, param);
            }
        }

        /// <summary>
        /// 所属的UIState，可以重载，不重载时表明该Mediator不属于任何状态
        /// </summary>
        /// <returns>uistate的数组</returns>
        protected virtual string[] GetBelongsUIStateName()
        {
            return null;
        }

        /// <summary>
        /// 当初始化时调用
        /// </summary>
        protected virtual void OnInitialize()
        {
        }

        /// <summary>
        /// 当销毁时调用调用
        /// </summary>
        protected virtual void OnUninitialize()
        {
        }

        /// <summary>
        /// 当UIState切入时调用
        /// </summary>
        /// <param name="changeType">切换类型</param>
        /// <param name="oldStateName">老状态名，可能为空和null</param>
        /// <param name="newStateName">新状态名，可能为空和null</param>
        /// <param name="stateParam">状态参数</param>
        protected virtual void OnUIStateIn(int changeType, string oldStateName, string newStateName, object stateParam)
        {
        }

        /// <summary>
        /// 当UIState切出时调用
        /// <param name="changeType">切换类型</param>
        /// <param name="oldStateName">老状态名，可能为空和null</param>
        /// <param name="newStateName">新状态名，可能为空和null</param>
        /// </summary>
        protected virtual void OnUIStateOut(int changeType, string oldStateName, string newStateName)
        {
        }

        /// <summary>
        /// 状态刷新时调用
        /// <param name="changeType">切换类型</param>
        /// <param name="oldStateName">老状态名，可能为空和null</param>
        /// <param name="newStateName">新状态名，可能为空和null</param>
        /// <param name="stateParam">状态参数</param>
        /// </summary>
        protected virtual void OnUIStateStay(int changeType, string oldStateName, string newStateName, object stateParam)
        {
        }


        /// <summary>
        /// 更新UI
        /// </summary>
        /// <param name="id">更新ID标识</param>
        /// <param name="param">更新参数</param>
        /// <returns>是否阻断向下传递刷新</returns>
        protected virtual bool OnUpdateUI(string id, object param)
        {
            return false;
        }

        /// <summary>
        /// 操作
        /// </summary>
        /// <param name="id">操作ID标识</param>
        /// <param name="param">操作参数</param>
        /// <returns>是否阻断向上传递操作</returns>
        protected virtual bool OnAction(string id, object param)
        {
            return false;
        }

        /// <summary>
        /// 获取是否在UIMediator所属的状态中
        /// </summary>
        /// <returns>是否在UIMediator所属的状态中</returns>
        protected bool IsInState()
        {
            return _inState;
        }

        /// <summary>
        /// 创建窗口对象
        /// </summary>
        /// <typeparam name="T">对象类型</typeparam>
        /// <param name="usePool">是否使用对象池</param>
        /// <param name="customID">自定义ID</param>
        /// <param name="parameter">预制件初始化参数</param>
        /// <returns>预制件对象</returns>
        protected T CreateFormClass<T>(bool usePool, int customID = 0, object parameter = null) where T : UIFormClass, new()
        {
            T prefabClass = UIFormHelper.CreateFormClass<T>(usePool, this, customID, parameter);
            if (prefabClass == null)
            {
                JW.Common.Log.LogE("UIMediator.CreatePrefabClass : failed to create {0}", typeof(T).FullName);
                return null;
            }

            _forms.Add(prefabClass);
            return prefabClass;
        }

        /// 销毁窗口对象
        protected void DisposeFormClass<T>(ref T prefabClass) where T : UIFormClass, new()
        {
            if (null == prefabClass)
            {
                return;
            }

            for (int i = 0; i < _forms.Count; i++)
            {
                if (_forms[i] == prefabClass)
                {
                    _forms[i] = null;
                    break;
                }
            }
            //todo 这里应该等Form FadeOut动画完才能销毁
            UIFormHelper.DisposeFormClass(ref prefabClass);
        }


        /// <summary>
        /// 删除所有的窗口
        /// </summary>
        protected void DestroyAllFormClass()
        {
            for (int i = 0; i < _forms.Count; i++)
            {
                UIFormClass pc = _forms[i];
                if (pc == null)
                {
                    continue;
                }
                _forms[i] = null;
                UIFormHelper.DisposeFormClass(pc);
            }
        }
    }
}

